File: taskpaper_plugin.rb
require 'json'

payload = JSON.parse($stdin.read)
text    = payload['text']

# Handle YAML frontmatter — convert to bold key: value display
if text.start_with?("---\n")
  fm_end = text.index("\n---\n", 4) || text.index("\n...\n", 4)
  if fm_end
    fm_content = text[4...fm_end]
    rest = text[(fm_end + 5)..]
    fm_display = fm_content.each_line.map do |line|
      if (m = line.match(/\A([^:\n]+):\s*(.*)\n?\z/))
        "**#{m[1]}:**{.fmkey} #{m[2]}  \n"
      else
        line
      end
    end.join
    text = "\\-\\-\\-\n" + fm_display + "\\-\\-\\-\n\n" + rest
  end
end

fence_indent = nil
fence_ticks  = nil

# Single pass: plain-text output (no list conversion)
# Project names get bold + color; tasks get escaped dash; @tags handled by atag plugin
result = text.each_line.map do |line|
  if fence_indent.nil?
    if (m = line.match(/\A(\s*)(`{3,})/))
      fence_indent = m[1].length
      fence_ticks  = m[2].length
      next line
    end
  else
    if (m = line.match(/\A(\s*)(`{3,})\s*\z/)) &&
        m[1].length <= fence_indent && m[2].length >= fence_ticks
      fence_indent = nil
      fence_ticks  = nil
    end
    next line
  end

  if (m = line.match(/\A(\t*)[^\s#>].*:\s*\z/)) && !line.match?(/https?:/)
    tabs   = m[1].length
    indent = "\u00A0\u00A0" * [tabs - 1, 0].max
    cls    = 'h4text'
    "#{indent}**#{line.sub(/\A\t*/, '').rstrip}**{.#{cls}}  \n"
  elsif line.match?(/\A\t+- /)
    tabs   = line.index('-')
    indent = "\u00A0\u00A0" * [tabs - 1, 0].max
    "#{indent}\\- #{line.sub(/\A\t+- /, '').rstrip}  \n"
  elsif line.match?(/\A {4}/)
    spaces = line[/\A */].length
    level  = spaces / 4
    rest   = line.sub(/\A +/, '')
    rest.start_with?("- ") \
      ? "#{"\u00A0\u00A0" * [level - 1, 0].max}\\- #{rest[2..].rstrip}  \n" \
      : line
  elsif (m = line.chomp.match(/\A(##?) (.+)\z/))
    hashes  = m[1]
    content = m[2].rstrip
    prefix  = '\\#' * hashes.length
    "#{hashes} **#{prefix} #{content}**\n"
  elsif (m = line.chomp.match(/\A(####?) (.+)\z/))
    prefix = '\\#' * m[1].length
    "**#{prefix} #{m[2].rstrip}**{.h3text}  \n"
  else
    line
  end
end.join

print result

File: taskpaper_plugin.go
package main

import (
	"encoding/json"
	"io"
	"os"
	"regexp"
	"strings"
)

var (
	reFenceOpen  = regexp.MustCompile("^([ \t]*)(`{3,})")
	reFenceClose = regexp.MustCompile("^([ \t]*)(`{3,})\\s*$")
	reProject    = regexp.MustCompile(`^(\t*)[^\s#>].*:\s*(@\w+(\([^)]*\))?\s*)*$`)
	reTabTask    = regexp.MustCompile(`^(\t+)- (.*)`)
	reH12        = regexp.MustCompile(`^(##?) (.+)$`)
	reH34        = regexp.MustCompile(`^(####?) (.+)$`)
	reFMLine     = regexp.MustCompile(`^([^:\n]+):\s*(.*)`)
	reMMDLine    = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9 _-]*:\s+\S`)
)

const doubleNbsp = "  "

func nbspIndent(n int) string {
	if n <= 0 {
		return ""
	}
	return strings.Repeat(doubleNbsp, n)
}

// splitLines splits text into lines preserving line endings.
func splitLines(text string) []string {
	if text == "" {
		return nil
	}
	var lines []string
	for text != "" {
		i := strings.IndexByte(text, '\n')
		if i < 0 {
			lines = append(lines, text)
			break
		}
		lines = append(lines, text[:i+1])
		text = text[i+1:]
	}
	return lines
}

func fmLineDisplay(line string) string {
	chomped := strings.TrimRight(line, "\n")
	if m := reFMLine.FindStringSubmatch(chomped); m != nil {
		return "**" + m[1] + ":**{.fmkey} " + m[2] + "  \n"
	}
	return line
}

func processFrontmatter(text string) string {
	if strings.HasPrefix(text, "---\n") {
		endPos := -1
		if i := strings.Index(text[4:], "\n---\n"); i >= 0 {
			endPos = 4 + i
		} else if i := strings.Index(text[4:], "\n...\n"); i >= 0 {
			endPos = 4 + i
		}
		if endPos >= 0 {
			var sb strings.Builder
			for _, line := range splitLines(text[4:endPos]) {
				sb.WriteString(fmLineDisplay(line))
			}
			return "\\-\\-\\-\n" + sb.String() + "\\-\\-\\-\n\n" + text[endPos+5:]
		}
	}

	lines := splitLines(text)
	mmdEnd := 0
	for mmdEnd < len(lines) && reMMDLine.MatchString(lines[mmdEnd]) {
		mmdEnd++
	}
	if mmdEnd > 0 && (mmdEnd >= len(lines) || strings.TrimRight(lines[mmdEnd], "\n") == "") {
		var sb strings.Builder
		for _, line := range lines[:mmdEnd] {
			sb.WriteString(fmLineDisplay(line))
		}
		restStart := mmdEnd
		if mmdEnd < len(lines) {
			restStart = mmdEnd + 1
		}
		return sb.String() + "\n" + strings.Join(lines[restStart:], "")
	}

	return text
}

func leadingSpaces(s string) int {
	for i := 0; i < len(s); i++ {
		if s[i] != ' ' {
			return i
		}
	}
	return len(s)
}

func processLines(text string) string {
	fenceIndent := -1
	fenceTicks := -1
	var sb strings.Builder

	for _, line := range splitLines(text) {
		chomped := strings.TrimRight(line, "\n")

		if fenceIndent < 0 {
			if m := reFenceOpen.FindStringSubmatch(chomped); m != nil {
				fenceIndent = len(m[1])
				fenceTicks = len(m[2])
				sb.WriteString(line)
				continue
			}
		} else {
			if m := reFenceClose.FindStringSubmatch(chomped); m != nil &&
				len(m[1]) <= fenceIndent && len(m[2]) >= fenceTicks {
				fenceIndent = -1
				fenceTicks = -1
			}
			sb.WriteString(line)
			continue
		}

		switch {
		case reProject.MatchString(chomped):
			m := reProject.FindStringSubmatch(chomped)
			tabs := len(m[1])
			noTabs := strings.TrimLeft(chomped, "\t")
			sb.WriteString(nbspIndent(max(tabs-1, 0)) + "**" + strings.TrimRight(noTabs, " \t") + "**{.h4text}  \n")

		case reTabTask.MatchString(chomped):
			m := reTabTask.FindStringSubmatch(chomped)
			tabs := len(m[1])
			sb.WriteString(nbspIndent(max(tabs-1, 0)) + "\\- " + strings.TrimRight(m[2], " \t") + "  \n")

		case strings.HasPrefix(chomped, "    "):
			spaces := leadingSpaces(chomped)
			level := spaces / 4
			rest := chomped[spaces:]
			if strings.HasPrefix(rest, "- ") {
				sb.WriteString(nbspIndent(max(level-1, 0)) + "\\- " + strings.TrimRight(rest[2:], " \t") + "  \n")
			} else {
				sb.WriteString(line)
			}

		case reH12.MatchString(chomped):
			m := reH12.FindStringSubmatch(chomped)
			hashes := m[1]
			content := strings.TrimRight(m[2], " \t")
			prefix := strings.Repeat("\\#", len(hashes))
			sb.WriteString(hashes + " **" + prefix + " " + content + "**\n")

		case reH34.MatchString(chomped):
			m := reH34.FindStringSubmatch(chomped)
			prefix := strings.Repeat("\\#", len(m[1]))
			sb.WriteString("**" + prefix + " " + strings.TrimRight(m[2], " \t") + "**{.h3text}  \n")

		default:
			sb.WriteString(line)
		}
	}

	return sb.String()
}

func main() {
	data, _ := io.ReadAll(os.Stdin)
	var payload map[string]json.RawMessage
	json.Unmarshal(data, &payload)
	var text string
	json.Unmarshal(payload["text"], &text)

	os.Stdout.WriteString(processLines(processFrontmatter(text)))
}
